### PR TITLE
feat: update connectors and style

### PR DESCRIPTION
**Pull request by cmaloney111**
- Adds upstream connectors to different document source (for now, Google drive, S3, URLs, and local directories).
- Unifies all functions related to parsing documents into one 'parse' function (but maintains backwards compatibility with previous functions)
- Updates tests for connectors and new tests for other edge cases
- Updates README to reflect changes

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -65,36 +65,116 @@ The library can extract data from:
...
+)
+
+# Parse all documents in the bucket
+results = parse(config)
+
+# Parse documents in a specific prefix/folder
+results = parse(config, connector_path="documents/")
+```
+
+##### Local Directory Connector
```
Comment (humpydonkey) [inline README.md:152]: it's helpful to clarify the local connector doesn't support nested folders in the doc.

**Review by humpydonkey — COMMENTED** [review]
Fantastic job on adding the tests, the codebase is much more covered with the PR! @cmaloney111

File: README.md (hunk 1 above)
Comment (cmaloney111) [reply README.md:152]: Just added a recursive option and extra info in the README

File: agentic_doc/connectors.py [hunk 2]
```diff
@@ -0,0 +1,550 @@
...
+                "*.png",
+                "*.jpg",
+                "*.jpeg",
+                "*.gif",
+                "*.bmp",
+                "*.tiff",
+            ]  # TODO: Include all supported patterns
+            files = []
+            for pat in patterns:
+                files.extend(search_path.glob(pat))
```
[the remark this answers is not in the document]
Comment (cmaloney111) [reply agentic_doc/connectors.py]: Made sure to add all image extensions and support for uppercase extensions. It should perform one pass now and not add unnecessary patterns. It also supports recursive searches with the recursive option passed in the config

File: agentic_doc/connectors.py [hunk 3]
```diff
@@ -0,0 +1,550 @@
...
+    @abstractmethod
+    def list_files(
+        self, path: Optional[str] = None, pattern: Optional[str] = None
+    ) -> List[str]:
+        """
+        List files available through this connector.
+
+        Args:
+            path: Optional path to list files from
+            pattern: Optional pattern to filter files
```
[the remark this answers is not in the document]
Comment (cmaloney111) [reply agentic_doc/connectors.py]: Added pathlib-like glob support for all connectors (except URL), and made that clear in the docs.

File: agentic_doc/connectors.py [hunk 4]
```diff
@@ -0,0 +1,550 @@
...
+        return []
+
+    def download_file(self, file_id: str, local_path: Optional[str] = None) -> Path:
+        """Download file from URL."""
+        try:
+            import httpx
+        except ImportError:
+            raise ImportError(
+                "URL connector requires httpx. Install with: pip install httpx"
+            )
```
[the remark this answers is not in the document]
Comment (cmaloney111) [reply agentic_doc/connectors.py]: 👍

File: agentic_doc/connectors.py [hunk 5]
```diff
@@ -0,0 +1,550 @@
...
+        """
+        pass
+
+    @abstractmethod
+    def download_file(self, file_id: str, local_path: Optional[str] = None) -> Path:
+        """
+        Download a file to local storage.
+
+        Args:
+            file_id: Identifier for the file to download
```
[the remark this answers is not in the document]
Comment (cmaloney111) [reply agentic_doc/connectors.py]: Done, added specifics for file_id for each connector so users know how to use this function. I also added a note that this function should generally be called on the result from list_files as opposed to trying to call this function alone

File: agentic_doc/connectors.py [hunk 6]
```diff
@@ -0,0 +1,550 @@
...
+                # Filter by file extension (documents and images)
+                valid_extensions = [
+                    ".pdf",
+                    ".png",
+                    ".jpg",
+                    ".jpeg",
+                    ".gif",
+                    ".bmp",
+                    ".tiff",
+                ]
```
[the remark this answers is not in the document]
Comment (cmaloney111) [reply agentic_doc/connectors.py]: Done

File: agentic_doc/connectors.py [hunk 7]
```diff
@@ -0,0 +1,550 @@
...
+            raise
+
+    def get_file_info(self, file_id: str) -> Dict[str, Any]:
+        """Get file info from URL headers."""
+        try:
+            import httpx
+        except ImportError:
+            raise ImportError(
+                "URL connector requires httpx. Install with: pip install httpx"
+            )
```
[the remark this answers is not in the document]
Comment (cmaloney111) [reply agentic_doc/connectors.py]: 👍

File: agentic_doc/connectors.py [hunk 8]
```diff
@@ -0,0 +1,550 @@
...
+        self.config: S3ConnectorConfig = config
+        self._client: Optional[S3Client] = None
+
+    def _get_client(self) -> S3Client:
+        """Initialize S3 client if not already done."""
+        if self._client is None:
+            if "boto3" not in sys.modules:
+                raise ImportError(
+                    "S3 connector requires boto3. Install with: pip install boto3"
+                )
```
[the remark this answers is not in the document]
Comment (cmaloney111) [reply agentic_doc/connectors.py]: 👍

File: agentic_doc/parse.py [hunk 9]
```diff
@@ -235,6 +388,7 @@ def _merge_part_results(results: list[ParsedDocument]) -> ParsedDocument:
             start_page_idx=0,
             end_page_idx=0,
             doc_type="pdf",
+            result_path=Path(),
```
Comment (AsiaCao) [inline agentic_doc/parse.py:391]: is it better to be None? consistent with line 375, `_parse_image`

**Review by AsiaCao — APPROVED** [review]
nice work!

**Comment by cmaloney111** [comment]
Thanks!
